contrib/epee/include/net/abstract_tcp_server.h

```diff
@@ -305,7 +305,7 @@ namespace net_utils
 		m_connections.back().powner = this;
 		m_connections.back().m_self_it = --m_connections.end();
 		m_connections.back().m_context.m_remote_address = remote_address;
-		m_connections.back().m_htread = threads_helper::create_thread(ConnectionHandlerProc, &m_connections.back());
+		m_connections.back().m_htread = threads_helper::create_thread(ConnectionHandlerProc, &m_connections.back()); // ugh, seems very risky
```

Comment:
Why is this risky? Can `m_connections` become empty at this point?

Reply:
If I'm reading it right, the m_connections object isn't guarded by a lock in this section of the code so it is entirely possible that another thread_context is added to m_connections at any point during lines 304-308 which would cause the back() calls to return different objects.

In addition, the ConnectionHandlerProc, in another thread, could potentially erase() the thread_context (line 154) prior to line 308 running.

Allocating a thread_context() from the heap would allow you to keep the pointer in the list and just use remove() instead of erase(). That way, you can keep the m_connections longer.

Using a condition variable in ConnectionHandlerProc and signaling before invoke_connection() returns would solve the second race.

Not entirely up to speed with c++containers, but that is what I see.